packages/sources/kaiko/src/adapter.ts

```diff
@@ -75,3 +82,71 @@ export const execute: ExecuteWithConfig<Config> = async (request, config) => {
 export const makeExecute: ExecuteFactory<Config> = (config) => {
   return async (request) => execute(request, config || makeConfig())
 }
+
+type MessageType = 'info' | 'error' | 'update'
+
+interface Message {
+  event: MessageType
+  payload: Record<string, any>
+}
+
+interface UpdateMessage extends Message {
+  event: 'info'
+  payload: {
+    subscription: {
+      topic: string
+      data_version: string
+      exchange: string
+      instrument_class: string
+      instrument: string
+    }
+    data: {
+      timestamp: number
+      trade_id: string
+      price: string
+      amount: string
+      taker_side_sell: boolean
+    }[]
+  }
+}
+
+export const makeWSHandler = (config?: Config): MakeWSHandler => {
+  const getSubscription = (instrumentCode?: string, subscribe = true) => {
+    if (!instrumentCode) return
+    return {
+      command: subscribe ? 'subscribe' : 'unsubscribe',
+      args: {
+        subscriptions: {
+          topic: 'trades_ws',
+          pattern: `*:spot:${instrumentCode}`,
+          // "data_version": <data_version> // Defaults to latest version
+        },
+      },
+    }
+  }
+  const getInstrument = (input: AdapterRequest) => {
+    const validator = new Validator(input, customParams)
+    if (validator.error) return
+    const base = validator.validated.data.base.toUpperCase()
+    const quote = validator.validated.data.quote.toUpperCase()
+    return `${base}-${quote}`
+  }
+  return () => {
+    const defaultConfig = config || makeConfig()
+    return {
+      connection: {
+        url: defaultConfig.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
```

Comment:
Is the authentication missing?

Reply:
Must be, I was having problems connecting.